tsdb/blockwriter.go

```diff
@@ -0,0 +1,132 @@
+)
+
+// Writer is interface to write time series into Prometheus blocks.
+type Writer interface {
+	storage.Appendable
+
+	// Flush writes current data to disk.
+	// The block or blocks will contain values accumulated by `Write`.
+	Flush(ctx context.Context) (ulid.ULID, error)
+
+	// Close releases all resources.
+	Close() error
+}
+
+var _ Writer = &TSDBWriter{}
+
+// Writer is a block writer that allows appending and flushing series to disk.
+type TSDBWriter struct {
+	logger log.Logger
+	dir    string
+
+	head   *Head
+	tmpDir string
+}
+
+func durationToMillis(t time.Duration) int64 {
+	return int64(t.Seconds() * 1000)
+}
+
+// NewTSDBWriter create a new block writer.
+//
+// The returned writer accumulates all the series in memory until `Flush` is called.
+//
+// Note that the writer will not check if the target directory exists or
+// contains anything at all. It is the caller's responsibility to
+// ensure that the resulting blocks do not overlap etc.
+// Writer ensures the block flush is atomic (via rename).
+func NewTSDBWriter(logger log.Logger, dir string) *TSDBWriter {
+	return &TSDBWriter{
+		logger: logger,
+		dir:    dir,
+	}
+}
+
+// initHead creates and initialises a new TSDB head.
+// blockSize should be in milliseconds.
+func (w *TSDBWriter) initHead(blockSize int64) error {
+	logger := w.logger
+
+	tmpDir, err := ioutil.TempDir(os.TempDir(), "head")
+	if err != nil {
+		return errors.Wrap(err, "create temp dir")
+	}
+	w.tmpDir = tmpDir
+
+	h, err := NewHead(nil, logger, nil, blockSize, w.tmpDir, nil, DefaultStripeSize, nil)
+	if err != nil {
+		return errors.Wrap(err, "tsdb.NewHead")
+	}
+
+	w.head = h
+	return w.head.Init(math.MinInt64)
+}
+
+// Appender returns a new appender on the database.
+// Appender can't be called concurrently. However, the returned Appender can safely be used concurrently.
+func (w *TSDBWriter) Appender(ctx context.Context) storage.Appender {
+	return w.head.Appender(ctx)
+}
+
+// Flush implements the Writer interface. This is where actual block writing
+// happens. After flush completes, no writes can be done.
+func (w *TSDBWriter) Flush(ctx context.Context) (ulid.ULID, error) {
+	seriesCount := w.head.NumSeries()
+	if w.head.NumSeries() == 0 {
+		return ulid.ULID{}, errors.New("no series appended; aborting.")
+	}
+
+	mint := w.head.MinTime()
+	maxt := w.head.MaxTime()
+	level.Info(w.logger).Log("msg", "flushing", "series_count", seriesCount, "mint", timestamp.Time(mint), "maxt", timestamp.Time(maxt))
+
+	compactor, err := NewLeveledCompactor(ctx,
+		nil,
+		w.logger,
+		[]int64{durationToMillis(2 * time.Hour)},
+		chunkenc.NewPool())
+	if err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "create leveled compactor")
+	}
+	id, err := compactor.Write(w.dir, w.head, mint, maxt, nil)
+	if err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "compactor write")
+	}
+
+	return id, nil
+}
+
+func (w *TSDBWriter) Close() error {
+	_ = os.RemoveAll(w.tmpDir)
```

Comment:
This should be in a `defer` because the head.Close() can/will have some file operations (like flushing some chunks to disk for the m-map which is in the buffer).